src/python/pants/rules/core/list_roots.py

```diff
@@ -0,0 +1,55 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from typing import Set
+
+from pants.engine.console import Console
+from pants.engine.fs import PathGlobs, Snapshot
+from pants.engine.goal import Goal, LineOriented
+from pants.engine.rules import console_rule, optionable_rule, rule
+from pants.engine.selectors import Get
+from pants.source.source_root import AllExistingSourceRoots, SourceRoot, SourceRootConfig
+
+
+class Roots(LineOriented, Goal):
+  """List the repo's registered source roots."""
+  name = 'roots'
+
+
+@rule(AllExistingSourceRoots, [SourceRootConfig])
+def all_roots(source_root_config):
+
+  source_roots = source_root_config.get_source_roots()
+
+  all_paths: Set[str] = set()
+  for path in source_roots.traverse():
+    if path.startswith("^/"):
+      all_paths |= {f"{path[2:]}/"}
+    else:
+      all_paths |= {f"**/{path}/"}
+
+  snapshot = yield Get(Snapshot, PathGlobs(include=tuple(all_paths)))
+
+  all_source_roots: Set[SourceRoot] = set()
+  for directory in sorted(snapshot.dirs):
+    match: SourceRoot = source_roots.trie_find(directory)
```

Comment:
What is an example of a case where a directory doesn't match in the trie? Didn't we construct all_paths so that they have to match?  If there is such a case then it needs to be documented here to explain why we do this extra match.

Reply:
The PathGlobs -> Snapshot rule is picking up subdirectories of the actual source roots (for instance picking up subdirectories of `src/rust/*`) that would get pruned out by calling `find_by_path()` (which is what should be here instead of `trie_find`)